infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java

```diff
@@ -12,33 +16,148 @@ public abstract class DefaultAdapterForUpgrade implements ILogReplicationVersion
+
+    /* Represents the current code's version
+     */
+    public static final String NODE_VERSION_KEY = "VERSION";
+
     String versionString = "version_latest";
+
+    String pinnedVersionString = "version_latest";
     CorfuStore corfuStore;
 
-    @Override
-    public String getVersion() {
-        if (corfuStore != null) {
-            try {
-                corfuStore.openTable(NAMESPACE, VERSION_TEST_TABLE,
-                        LogReplicationStreams.VersionString.class,
-                        LogReplicationStreams.Version.class, CommonTypes.Uuid.class,
-                        TableOptions.builder().build());
-            } catch (Exception e) {
-                // Just for wrap this up
-            }
+    public DefaultAdapterForUpgrade(CorfuRuntime runtime) {
+        this.corfuStore = new CorfuStore(runtime);
+        openVersionTable(corfuStore);
+    }
 
-            LogReplicationStreams.VersionString versionStringKey =
-                    LogReplicationStreams.VersionString.newBuilder()
-                            .setName("VERSION").build();
+    @Override
+    public String getNodeVersion() {
+        // For local testing we could be either within a transaction..
+        if (TransactionalContext.isInTransaction()) {
+            getVersion(TransactionalContext.getRootContext().getTxnContext());
+        } else { // or invoking this independently outside of a transaction..
             try (TxnContext txn = corfuStore.txn(NAMESPACE)) {
-                versionString =
-                        ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
-                                versionStringKey).getPayload()).getVersion();
+                getVersion(txn);
                 txn.commit();
             } catch (Exception e) {
-                // Just for wrap this up
+                // Just to wrap this up
             }
         }
         return versionString;
     }
+
+    private void getVersion(TxnContext txn) {
+        LogReplicationStreams.VersionString versionStringKey =
+                LogReplicationStreams.VersionString.newBuilder()
+                        .setName(NODE_VERSION_KEY).build();
+        versionString =
+                ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
+                        versionStringKey).getPayload()).getVersion();
+    }
+
+    private void getPinnedVersionString(TxnContext txn) {
+        LogReplicationStreams.VersionString versionStringKey =
+                LogReplicationStreams.VersionString.newBuilder()
+                        .setName(PINNED_VERSION_KEY).build();
+        pinnedVersionString =
+                ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
+                        versionStringKey).getPayload()).getVersion();
+    }
+
+    @Override
+    public String getPinnedClusterVersion(TxnContext txnContext) {
+        getPinnedVersionString(txnContext);
+        return pinnedVersionString;
+    }
```

Comment:
why we create a separate overloaded method here? seems like the pinnedVersionString could be directly created here?

Reply:
the `getPinnedClusterVersion` is coming from the interface now.